Index: inc/classes/generate.class.php
===================================================================
--- inc/classes/generate.class.php	(revision 1306741)
+++ inc/classes/generate.class.php	(working copy)
@@ -1461,7 +1461,9 @@
 	 * 			@param object $post The Post Object.
 	 * 			@param bool $external Wether to fetch the current WP Request or get the permalink by Post Object.
 	 * 			@param bool $is_term Fetch url for term.
-	 * 			@param bool $term The term object.
+	 * 			@param object $term The term object.
+	 * 			@param bool $home Fetch home URL.
+	 * 			@param bool $forceslash Fetch home URL and slash it, always.
 	 * }
 	 *
 	 * @since 2.0.0
@@ -1644,17 +1646,21 @@
 		if ( ! isset( $scheme ) )
 			$scheme = is_ssl() ? 'https' : 'http';
 
+		$output = $this->set_url_scheme( $url, $scheme );
+
 		/**
 		 * Slash it only if $slashit is true
 		 *
 		 * @since 2.2.4
 		 */
-		if ( $slashit ) {
-			$output = esc_url( user_trailingslashit( $this->set_url_scheme( $url, $scheme ) ) );
-		} else {
-			$output = esc_url( $this->set_url_scheme( $url, $scheme ) );
-		}
+		if ( $slashit && ! $args['forceslash'] )
+			$output = user_trailingslashit( $output );
 
+		if ( $args['forceslash'] )
+			$output = trailingslashit( $output );
+
+		$output = esc_url( $output );
+
 		/**
 		 * Debug parameters.
 		 * @since 2.4.2
@@ -1700,7 +1706,8 @@
 				'is_term' 			=> false,
 				'post' 				=> null,
 				'term'				=> null,
-				'home'				=> false
+				'home'				=> false,
+				'forceslash'		=> false
 			);
 		}
 
@@ -1717,6 +1724,7 @@
 		$args['post'] 				= isset( $args['post'] ) 				? (object) $args['post'] 			: $defaults['post'];
 		$args['term'] 				= isset( $args['term'] ) 				? (object) $args['term'] 			: $defaults['term'];
 		$args['home'] 				= isset( $args['home'] ) 				? (bool) $args['home'] 				: $defaults['home'];
+		$args['forceslash'] 		= isset( $args['forceslash'] ) 			? (bool) $args['forceslash'] 		: $defaults['forceslash'];
 
 		return $args;
 	}
Index: inc/classes/metaboxes.class.php
===================================================================
--- inc/classes/metaboxes.class.php	(revision 1306741)
+++ inc/classes/metaboxes.class.php	(working copy)
@@ -1635,7 +1635,7 @@
 
 		$site_url = $this->the_url_from_cache( '', '0', false, false );
 
-		$sitemap_url = esc_url( $site_url . 'sitemap.xml' );
+		$sitemap_url = esc_url( trailingslashit( $site_url ) . 'sitemap.xml' );
 
 		?>
 		<h4><?php _e( 'Sitemap Integration Settings', 'autodescription' ); ?></h4>
@@ -1672,7 +1672,7 @@
 
 		$site_url = $this->the_url_from_cache( '', '0', false, false );
 
-		$robots_url = esc_url( $site_url . 'robots.txt' );
+		$robots_url = esc_url( trailingslashit( $site_url ) . 'robots.txt' );
 		$here =  '<a href="' . $robots_url  . '" target="_blank" title="' . __( 'View robots.txt', 'autodescription' ) . '">' . _x( 'here', 'The sitemap can be found %s.', 'autodescription' ) . '</a>';
 
 		?>
Index: inc/classes/render.class.php
===================================================================
--- inc/classes/render.class.php	(revision 1306741)
+++ inc/classes/render.class.php	(working copy)
@@ -795,7 +795,7 @@
 	public function ld_json() {
 
 		//* Check for WPSEO LD+JSON
-		if ( $this->has_json_ld_plugin() !== false )
+		if ( $this->has_json_ld_plugin() !== false || is_search() || is_404() )
 			return;
 
 		$this->setup_ld_json_transient( get_the_ID() );
Index: inc/classes/search.class.php
===================================================================
--- inc/classes/search.class.php	(revision 1306741)
+++ inc/classes/search.class.php	(working copy)
@@ -53,7 +53,7 @@
 
 		$cache_key = 'exclude_search_ids_' . $blog_id;
 
-		$post_ids = $this->ojbect_cache_get( $cache_key );
+		$post_ids = $this->object_cache_get( $cache_key );
 		if ( false === $post_ids ) {
 			$post_ids = array();
 
Index: inc/classes/transients.class.php
===================================================================
--- inc/classes/transients.class.php	(revision 1306741)
+++ inc/classes/transients.class.php	(working copy)
@@ -154,6 +154,8 @@
 	 *
 	 * @staticvar array $cached_id
 	 *
+	 * @global $blog_id;
+	 *
 	 * @since 2.3.3
 	 *
 	 * @return string The generated page id key.
@@ -167,76 +169,118 @@
 
 		global $blog_id;
 
-		if ( ( $page_id === false || is_front_page() ) && ( 'posts' == get_option( 'show_on_front' ) ) ) {
-			if ( is_404() ) {
-				$page_id = '_404_';
-			} else {
+		$the_id = '';
+
+		/**
+		 * Generate home page cache key for the Home Page metabox.
+		 * @since 2.4.3.1
+		 */
+		if ( $this->is_menu_page( $this->pagehook ) ) {
+			if ( 'posts' == get_option( 'show_on_front' ) ) {
 				/**
-				 * Generate for home is blog page.
-				 * New transient name because of the conflicting bugfix on blog.
+				 * Detected home page.
 				 * @since 2.3.4
 				 */
-				$page_id = 'hblog_' . (string) get_option( 'page_on_front' );
-			}
-		} else if ( ( $page_id === false || is_front_page() || $page_id == get_option( 'page_on_front' ) ) && ( 'page' == get_option( 'show_on_front' ) ) ) {
-			if ( is_404() ) {
-				$page_id = '_404_';
+				$the_id = 'hpage_' . (string) get_option( 'page_on_front' );
 			} else {
 				/**
 				 * Detected home page.
 				 * @since 2.3.4
 				 */
-				$page_id = 'hpage_' . (string) get_option( 'page_on_front' );
+				$the_id = 'hpage_' . (string) get_option( 'page_on_front' );
 			}
-		} else if ( !is_front_page() && empty( $taxonomy ) && ( ( $page_id == get_option( 'page_for_posts' ) && get_option( 'page_for_posts' ) != 0 ) || ( $page_id === false && did_action( 'admin_init' ) ) ) ) {
-			/**
-			 * Generate key for blog page that's not the home page.
-			 * Bugfix
-			 * @since 2.3.4
-			 */
-			$page_id = 'blog_' . $page_id;
-		} else if ( !is_singular() && empty( $taxonomy ) && ! did_action( 'admin_init' ) ) {
-			global $wp_query;
+		} else {
+			if ( ! is_search() ) {
+				if ( ( $page_id === false || is_front_page() ) && ( 'posts' == get_option( 'show_on_front' ) ) ) {
+					if ( is_404() ) {
+						$the_id = '_404_';
+					} else {
+						/**
+						 * Generate for home is blog page.
+						 * New transient name because of the conflicting bugfix on blog.
+						 * @since 2.3.4
+						 */
+						$the_id = 'hblog_' . (string) get_option( 'page_on_front' );
+					}
+				} else if ( ( $page_id === false || is_front_page() || $page_id == get_option( 'page_on_front' ) ) && ( 'page' == get_option( 'show_on_front' ) ) ) {
+					if ( is_404() ) {
+						$the_id = '_404_';
+					} else {
+						/**
+						 * Detected home page.
+						 * @since 2.3.4
+						 */
+						$the_id = 'hpage_' . (string) get_option( 'page_on_front' );
+					}
+				} else if ( ! is_front_page() && empty( $taxonomy ) && ( ( $page_id == get_option( 'page_for_posts' ) && get_option( 'page_for_posts' ) != 0 ) || ( $page_id === false && did_action( 'admin_init' ) ) ) ) {
+					/**
+					 * Generate key for blog page that's not the home page.
+					 * Bugfix
+					 * @since 2.3.4
+					 */
+					$the_id = 'blog_' . $page_id;
+				} else if ( ! is_singular() && empty( $taxonomy ) && ! did_action( 'admin_init' ) ) {
+					global $wp_query;
 
-			/**
-			 * Generate for everything else.
-			 * Doesn't work on admin_init action.
-			 */
+					/**
+					 * Generate for everything else.
+					 * Doesn't work on admin_init action.
+					 */
 
-			$query = isset( $wp_query->query ) ? (array) $wp_query->query : null;
+					$query = isset( $wp_query->query ) ? (array) $wp_query->query : null;
 
-			/**
-			 * Automatically generate transient based on query.
-			 *
-			 * Adjusted to comply with the 45 char limit.
-			 * @since 2.3.4
-			 */
-			if ( isset( $query ) ) {
-				$page_id = '';
+					/**
+					 * Automatically generate transient based on query.
+					 *
+					 * Adjusted to comply with the 45 char limit.
+					 * @since 2.3.4
+					 */
+					if ( isset( $query ) ) {
+						$the_id = '';
 
-				// Trim key to 2 chars.
-				foreach ( $query as $key => $value )
-					$page_id .= substr( $key, 0, 2 ) . '_' . mb_substr( $value, 0, 2 ) . '_' . get_queried_object_id() . '_';
+						// Trim key to 2 chars.
+						foreach ( $query as $key => $value )
+							$the_id .= substr( $key, 0, 2 ) . '_' . mb_substr( $value, 0, 2 ) . '_' . get_queried_object_id() . '_';
 
-				//* Remove final underscore
-				$page_id = rtrim( $page_id, '_' );
-			}
-		} else if ( !is_singular() && !empty( $taxonomy ) ) {
+						//* Remove final underscore
+						$the_id = rtrim( $the_id, '_' );
+					}
+				} else if ( ! is_singular() && ! empty( $taxonomy ) ) {
 
-			//* Strip the ID from the taxonomy name.
-			$tax_len = mb_strlen( $taxonomy );
-			$_id = mb_substr( $page_id, $tax_len );
+					//* Strip the ID from the taxonomy name.
+					$tax_len = mb_strlen( $taxonomy );
+					$_id = mb_substr( $page_id, $tax_len );
 
-			//* Empty page id.
-			$page_id = '';
+					//* Empty page id.
+					$the_id = '';
 
-			//* Save taxonomy name and split into words with 3 length.
-			$taxonomy_name = explode( '_', $taxonomy );
-			foreach ( $taxonomy_name as $name )
-				$page_id .= substr( $name, 0, 3 ) . '_';
+					//* Save taxonomy name and split into words with 3 length.
+					$taxonomy_name = explode( '_', $taxonomy );
+					foreach ( $taxonomy_name as $name )
+						$the_id .= substr( $name, 0, 3 ) . '_';
 
-			//* Put it all together.
-			$page_id = $page_id . ltrim( $_id, '_' );
+					//* Put it all together.
+					$the_id = $the_id . ltrim( $_id, '_' );
+				} else if ( ! empty( $page_id ) ) {
+					$the_id = $page_id;
+				}
+			} else {
+				//* Search query.
+				$query = '';
+
+				if ( function_exists( 'get_search_query' ) ) {
+					$search_query = get_search_query();
+
+					if ( !empty( $search_query ) )
+						$query = str_replace( ' ', '', $search_query );
+
+					//* Limit to 10 chars.
+					if ( mb_strlen( $query ) > 10 )
+						$query = mb_substr( $query, 0, 10 );
+				}
+
+				$the_id = $page_id . '_s_' . $query;
+			}
 		}
 
 		/**
@@ -243,17 +287,16 @@
 		 * Static Front page isn't set. Causes all kinds of problems :(
 		 * Noob. :D
 		 */
-		if ( empty( $page_id ) ) {
-			$page_id = 'home_noob';
+		if ( empty( $the_id ) ) {
+			$the_id = 'home_noob';
 		}
 
 		/**
-		 * This should be at most 25 chars. Unless the $blog_id is higher than 99999999.
+		 * This should be at most 25 chars. Unless the $blog_id is higher than 99,999,999.
 		 * Then some cache keys will conflict on every 10th blog ID from eachother which post something on the same day..
-		 * On the day archive.
-		 * With the same description setting (short).
+		 * On the day archive. With the same description setting (short).
 		 */
-		return $cached_id[$page_id][$taxonomy] = (string) $page_id . '_' . (string) $blog_id;
+		return $cached_id[$page_id][$taxonomy] = (string) $the_id . '_' . (string) $blog_id;
 	}
 
 	/**
